[RNMobile] Scroll to selected block that goes off the screen

jd-alexander opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/735

Gutenberg Mobile PR https://github.com/wordpress-mobile/gutenberg-mobile/pull/2771
## Description
<!-- Please describe what you have changed or added -->
This solution seeks to address the issue where a block may get cut off due to it being moved off the edge of the screen. In a nutshell, whenever the selected `BlockListItem` component (which is a parent component for rendered blocks) is updated then we check to see if this list item is within the current viewport and if not we are utilizing the underlying scroll behavior for the respective Lists on each platform to reposition that item on the screen.  

## How has this been tested?
To test follow the instructions below. 

1. Open a post. 
2. Select a block and press the **move** down button until the block reaches the edge of the screen. 
3. Observe that once the **Selected** block that is being moved is positioned outside the bounds of the screen then the editor automatically scrolls to that position. 
4. If you select a block and press the **move** up button, then the same behavior as expounded on above will take place. The editor will scroll instantaneously so that the block is in the viewport. 

## Screenshots <!-- if applicable -->
Below are two gifs of the functionality in action on Android and iOS

### Moving a block off screen will auto-scroll the editor on Android

<kbd><img src="https://user-images.githubusercontent.com/1509205/97938345-309ab800-1d4f-11eb-98af-9a4ca895c5d8.gif" width="320"></kbd>

### Moving a block off screen will auto-scroll the editor on iOS

<kbd><img src="https://user-images.githubusercontent.com/1509205/97938354-36909900-1d4f-11eb-9eeb-50b62917b8bc.gif" width="320"></kbd>

## Types of changes

The main tenets of the solution are as follows : 

- The `measure` method of the `View` component is being utilized to determine the width, height, and position of the view within the current page/screen. 
- The `measureLayout` method of the `View` component detects where within the `FlatList` the current `BlockListItem` exists so that we can utilize it's `y` position for scrolling behavior. 
- For scrolls, the implementation differs per platform because the `FlatList`'s scroll functionality on iOS is managed by the `KeyboardAwareFlatList` so the ref to it's underlying `ScrollView` was utilized while on Android, scrolling was accomplished using the `FlatList` component itself.  
- To detect if a `BlockListItem` is within the window/screen we check to see if the `pageY` value is below zero ( which means we would need to scroll up) or above the window height ( which means we would need to scroll down)
- You will notice within the code that we are utilizing `scrollAdjustmentY` in the scroll algorithm. This is to give the `BlockListItem` the ability to initiate scrolling without it's bottom or top being at the edge of the screen. This is also used within the calculation of the offset to scroll to so that we don't have an issue where we scroll to the bottom or top of a view and there's no space above or below the block.  

## Concerns

1. On Android, I have noticed that scrolling a large block will sometimes cause a bit of the block to be cut off. I will need to have another look at the algorithm in place to ensure that it incorporates the height of the block in the right way so this never happens. 
2. I realized that scrolling a block pass a really large block that takes up the entire screen messes with its scrolling. It's like it stops scrolling on that very page because the block was moved a page away. 

Below you will observe that the list scrolls but it stops on the screen with the block that has taken up the entire viewport. 

<kbd><img src="https://user-images.githubusercontent.com/1509205/97938887-976ca100-1d50-11eb-8070-10d01c645a69.gif" width="320"></kbd>

You will also observe the same behavior when trying to scroll up with the block up movement action. The block is not scrolled to. The list stops the scroll action on the part of the page with the larger block. 

<kbd><img src="https://user-images.githubusercontent.com/1509205/97938893-9b98be80-1d50-11eb-993f-0cc1c07fec9f.gif" width="320"></kbd>

I am wondering if this is being caused because a `FlatList` is `Virtualized` so the scroll is faster than the fill rate of the content being rendered asynchronously off-screen. If so, then probably a delay could be introduced to ensure the list has rendered or we would have to somehow extend the off-screen limit in some way. Another approach would be to have some form of timeout/flag mechanism that runs a check to see if the block is within the viewport after the scroll has completed and if not then it does another scroll to the position it was supposed to scroll to in the first place.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

jd-alexander committed:
added the ability to access the FlatList via a ref

jd-alexander committed:
added functionality to detect when a block isn't in the view port

jd-alexander committed:
added more ref logic and ability to scroll based on focus.

jd-alexander committed:
added scroll logic to each block and now the flat list ties into that.

jd-alexander committed:
created scroll implementation

jd-alexander committed:
Merge branch 'master' into rnmobile/issue-scroll

jd-alexander committed:
Simplified scrolling logic and removed unneeded code.

jd-alexander committed:
removed scrolling logic out of block.native.js

jd-alexander committed:
sync Podfile

jd-alexander committed:
put the scrolling logic into the block list item.

jd-alexander committed:
utilize the scrollViewRef for iOS and the actual FlatList for Android

jd-alexander committed:
added scroll y offset/adjustment for scroll behavior.

jd-alexander committed:
put the scroll adjustment in a variable.

jd-alexander committed:
named the value for the scroll offset

jd-alexander committed:
removed unneeded props.

jd-alexander committed:
added comments to the scrolling algorithm.

jd-alexander commented on packages/block-editor/src/components/block-list/block-list-item.native.js:
```diff
@@ -89,6 +93,52 @@ export class BlockListItem extends Component {
+
+			// this adjustment factor is utilized to create an invisible bounds at the top and bottom of the screen where once the View enters within
+			//any of these bounds then the scrolling action is triggered.
+			const scrollAdjustmentY = isIOS ? 200 : 100;
```
I am always concerned about hardcoding these types of values because I am not sure if they will support every screen size. I haven't done a thorough test on multiple screen sizes as yet as I wanted to get some opinions on the proposed solution here.

jd-alexander replied:
I ended up going for a more dynamic solution using the window's height. [Done here ](https://github.com/WordPress/gutenberg/pull/26596/commits/29a0ed129210e962bc43643543867a950fed2c80#diff-d2ed3f81be347a76e4b1ce3e9ddf8723b0bb8a532207ab5a46e9680912373648R118-R121)

jd-alexander committed:
added comments to the list scrolling function.

jd-alexander committed:
added comments and utilized screen height for adjustment factor

jd-alexander committed:
removed the ios / android specific factor and used one factor.

jd-alexander committed:
simplified comment.

jd-alexander committed:
fixed comment about scroll view

guarani commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -92,6 +112,10 @@ export class BlockList extends Component {
 		this.scrollViewRef = ref;
 	}
 
+	setListRef( ref ) {
+		this.listRef = ref;
+	}
+
```
Is this being used?

jd-alexander replied:
It was being used before but not anymore. I will remove it 🙏

guarani commented on packages/components/src/mobile/keyboard-aware-flat-list/index.android.js:
```diff
@@ -2,17 +2,18 @@
  * External dependencies
  */
 import { FlatList } from 'react-native';
-
```
Not sure if this line being removed was intentional, the codebase seems to use an empty line in these situations (it would be nice though to have a linter rule to ensure consistency, but that's another topic).

jd-alexander replied:
Thanks for spotting. I will remove it.

guarani commented on packages/components/src/mobile/keyboard-aware-flat-list/index.android.js:
```diff
@@ -2,17 +2,18 @@
-		<FlatList { ...props } />
-	</KeyboardAvoidingView>
-);
+export const KeyboardAwareFlatList = ( props ) => {
+	return (
+		<KeyboardAvoidingView style={ { flex: 1 } }>
+			<FlatList ref={ props.listRef } { ...props } />
+		</KeyboardAvoidingView>
+	);
+};
```
I'm curious, what was the reason for changing the lines surrounding `<FlatList ref={ props.listRef } { ...props } />`? It seems the other lines changed, but are not directly relevant to the PR. Was it due to a formatting change that the editor did on save?

jd-alexander replied:
Yes, it seems the editor probably formatted this. I will revert this and I will be mindful of it in other PRs 👍

guarani replied:
Makes sense, I've had some formatting issues with the editor too.

guarani commented on packages/block-editor/src/components/block-list/block-list-item.native.js:
```diff
@@ -89,6 +93,50 @@ export class BlockListItem extends Component {
+						( _x, _y, blockWidth, blockHeight, px, py ) => {
+							const window = Dimensions.get( 'window' );
+
+							if ( isSelected ) {
```
It looks like this can be moved up to the beginning of `scrollToSelectedBlockIfItsNotVisible` to a) avoid execution reaching here if there's no selected block and b) reduce the number of levels of indentation.
```

  scrollToSelectedBlockIfItsNotVisible() {
    const { isSelected } = this.props;
    if (!isSelected) return;
    ...
    
```

guarani commented on packages/block-editor/src/components/block-list/block-list-item.native.js:
```diff
@@ -89,6 +93,50 @@ export class BlockListItem extends Component {
+	scrollToSelectedBlockIfItsNotVisible() {
+		const { isSelected } = this.props;
+
+		if ( this.blockRef.current ) {
```
Just a nitpick: this could be changed to `if ( !this.blockRef.current ) return;` to reduce the number of levels of indentation.

jd-alexander replied:
Thanks for this. I did it this way because that's the way how I wrote code in Kotlin in WPAndroid and the linter would throw an error if I did this without the braces 😅

guarani reviewed (commented):
This looks really good! I think the approach is very solid too. 

I noticed this strange glitch on iOS, it's hard to see what's happening in the gif below, so here's the steps to reproduce:
1. Scroll down to the bottom of the gallery block in the demo content
2. Tap the bottom of the gallery block
3. Notice the editor jumps up to near the top of the content

I noticed you talked about issues with really large blocks, so maybe this is related. I couldn't reproduce with the current WPiOS build, so it looks to be related to these changes. Can you reproduce it?


| <img src="https://user-images.githubusercontent.com/1898325/98151274-aa958300-1eae-11eb-8cb3-8dead22cf1b9.gif" alt="Screen recording showing glitch on iOS" width="350">  |
| - |

jd-alexander commented:
> This looks really good! I think the approach is very solid too.

Thanks for the review @guarani 🙇 

> I noticed you talked about issues with really large blocks, so maybe this is related. I couldn't reproduce with the current WPiOS build, so it looks to be related to these changes. Can you reproduce it?

Yes, this issue the same as the one I described in the PR. I am going to see what I can do to resolve it. 

Do you have any thoughts about this comment : 

> I am wondering if this is being caused because a FlatList is Virtualized so the scroll is faster than the fill rate of the content being rendered asynchronously off-screen. If so, then probably a delay could be introduced to ensure the list has rendered or we would have to somehow extend the off-screen limit in some way. Another approach would be to have some form of timeout/flag mechanism that runs a check to see if the block is within the viewport after the scroll has completed and if not then it does another scroll to the position it was supposed to scroll to in the first place.

I am wondering how we can determine how the block list `FlatList` gets loaded ahead of time i.e how many blocks exist beyond the edge of the screen because it's probably not able to scroll to it because it's not fully rendered yet. I am really not sure. I will try to run some experiments to figure this out.

guarani commented:
One option could be to create a separate RN project and try to reproduce this behavior with a scrollable list of "blocks", that could help with determining if the issue is related to the virtualization of lists or something else.

jd-alexander commented:
> One option could be to create a separate RN project and try to reproduce this behavior with a scrollable list of "blocks", that could help with determining if the issue is related to the virtualization of lists or something else.

Thanks, Paul that's a really good idea! Locally, I came up with a fix that would start the scroll again if the block isn't actually on the screen after a delay. So it would scroll to the next page, realize the block isn't there and then scroll to the actual block position. ( This is necessary because it seems the Flat List only loads items for one screen ahead - I haven't investigated this as yet but these assumptions are what led to the solution I'm mentioning now.) 

 So far, I am thinking this might not be an optimal solution but in the interest of HACK week, I was wondering if it would be a sensible improvement because currently the block just goes off the edge which isn't good. Also, most users will not have blocks taking up the entire screen so it might be a justified compromise if no other solution is found. Wdyt?

jd-alexander commented:
The only I found with my local solution/fix for this large black problem is that sometimes when the scroll is completely the block would be close to the end of that screen and a third scroll would be started so I need to modify the scroll logic a bit to ensure that the `yOffset` that's generated does not go close to the edge of the screen else the block will think it's outside the bounds of the visible area which it isn't and initiate an unneeded scroll operation.

guarani commented:
> Locally, I came up with a fix that would start the scroll again if the block isn't actually on the screen after a delay. So it would scroll to the next page, realize the block isn't there and then scroll to the actual block position.

I could take a look if you'd like, feel free to push the commit up.

> So far, I am thinking this might not be an optimal solution but in the interest of HACK week, I was wondering if it would be a sensible improvement because currently the block just goes off the edge which isn't good. Also, most users will not have blocks taking up the entire screen so it might be a justified compromise if no other solution is found. Wdyt?

Yeah, I agree that most users won't have really long blocks. Is your preference to go forward with the latest code that you have on this PR (the code that I looked at already), or would you like to add the changes from your local test and I could take a look at that? I also agree that even if the solution is not perfect, if it's a good improvement we can ship it.

jd-alexander commented:
> I could take a look if you'd like, feel free to push the commit up.

> Yeah, I agree that most users won't have really long blocks. Is your preference to go forward with the latest code that you have on this PR (the code that I looked at already), or would you like to add the changes from your local test and I could take a look at that? I also agree that even if the solution is not perfect, if it's a good improvement we can ship it.

Thanks for the feedback Paul! The solution I have isn't finished as yet but I am going to see if I can add in a final piece soon once I wrap up some other tasks on my list so we can ship this because it's a good improvement so far. It's buggy so I prefer to iron them out before pushing so you can be of more help 😄

jd-alexander committed:
fixed a typo

jd-alexander committed:
An attempt fix where a scrolling bug is present with large blocks.

This current solution involves checking if the py value of the yOffset passed will be within the bounds of any of the off screen bounds that's defined and if so then the yOffset being used to scroll needs to be readjusted to be within those bounds so another scroll isn't triggered. We are utilizing the height of the Flat List container to determine how the scroll view is divided based on the list's height.

jd-alexander commented:
Hi @guarani just an update here. I pushed some code that I was working on for the issue but it doesn't work as well I intended it to. I am going to revisit this task at a later time and once I am able to get it to a good place that doesn't have any bugs and at that point, I will ping you for a follow-up review. Thanks for your help so far. 🙇

guarani commented:
> Hi @guarani just an update here. I pushed some code that I was working on for the issue but it doesn't work as well I intended it to. I am going to revisit this task at a later time and once I am able to get it to a good place that doesn't have any bugs, at that point I will ping you for a follow-up review. Thanks for your help so far. 🙇

Sure, no worries! Anytime
